### ISSUE TITLE
Is this abandoned?

### ISSUE THREAD

**Issue by herdibintang** [pmros/cyclow#5]
No commit since April

**Comment by mangr3n** [comment]
So it is abandoned, kind of. 

I'm forking the project to keep using it and expanding it.

This API is important and I'm deeply disappointed that it hasn't been continued.

**Comment by brucou** [comment]
@mangr3n yes the project looks abandoned. I followed it as I was working in parallel with a similar set of ideas. but with what I believed to be a better API, also inspired from the electronic circuit metaphor. Quickly for what I remember, the objectives were :  

- generic enough to build higher abstractions on top
- hierarchical component model
- ability to trace the dataflow (i.e. structured logging)
- visual representation of the dataflow (static graph)
- devtool (re-playing the log of events and visualize the program step by step)
- testability

I wrote that a long-time ago (a year and some, maybe two), I revisited it 4 months ago, and moved it to a [`breadboard` repo](https://github.com/brucou/breadboard). Then I got busy with [a componentization model for `cyclejs`](https://brucou.github.io/posts/a-componentization-framework-for-cyclejs/) for which I am currently giving the final touch before the first usable release.

I am interested in reviewing and pursuing the ideas exposed in breadboard (I learnt a lot since), which are similar, though more generic than `cyclow`. That was a bit further in my priority list, but if you have interest in a collaboration, I could put it as a higher priority. In the best case, we could work together on a conceptual design and an API (**not** an implementation, this is always the last phase for me). I forewarn however than the API I am seeking is higher-level than cyclow. In the worse case, would you be willing, I can give you some feedback on what you will be doing, taken from my experience researching the subject.

As a teaser, quickly summarizing the ideas (based on my failing memory) behind `breadboard` : 

- architecture based on integrated circuit metaphor : chips, circuits, and controller, motherboard
- chips are entities with input ports, output ports, a `transform` function mapping input to output and which is not necessary a pure function i.e. can perform effects
- chips have a unique serie number and a prototype model (instance obtained by settings parametrization)
- chips have readouts which allow to trace their input and output
- chips incorporate a simlulation port, from which the chip can receive test input and observed on the output
- a circuit is a set of chips, connected by links
- circuit have input ports and output ports exposed to the circuit user, which may differ from the inner wiring exposed to the inner chips (i.e. there is some interface adaptation between external interface and the inner interface)
- circuits have readouts which allow to trace their input and output
- circuits incorporate a simlulation port, from which the circuit can receive test input and observed on the output
- links are connectors between input ports and output ports, or vice versa
- a controller monitor and manage the dependent circuits ; initialization, removal, plugging new circuits
- all circuits and chips are automatically connected to the controller on the motherboard and communicate through commands
- the motherboard is the framework which produces the tracing tools, devtool, etc.

This is what I can gather from the documentation I wrote in its days, but there is more to be said or explained. This was just to give you an idea.

What do you think?

**Comment by mangr3n** [comment]
So, my thought based on a quick look is that I'm a fan of Rich Hickey's talk Simple Made Easy.

The point of the wiring  model is that it allows us to think in terms of signals on ports treating the internals like a black box.   The description above seems like complexity for the sake of looking like hardware models which are useful for building hardware.  I'm only interested in the model for the simplicity offered by thinking about composable reactive components.  Reducing the connections to a signal graph is a very powerful paradigm.  Adding more to it than that simplicity feels like adding incidental complexity, not inherent complexity. 

So my question is: what does the complexity that you described above give you that graflow doesn't?

**Comment by brucou** [comment]
I know the talk that you are referring to, it is a pretty good one.

My description is pretty summarized, and is a rehash of the types I wrote in its days. A better presentation with illustrating figure and additional explanation would help for sure. However, do not confuse the metaphor with the intent. The metaphor is a way to introduce it to the world in an amenable way. The intent is to divide an application into smaller self-contained units, each of which of complexity manageable enough to be easy to write and reason about, and glue those smaller parts into the target application under development. What I propose is basically a componentization model attached to a communication model, and it is way more generic that cyclow. 

Genericity is not complexity. In fact there are two ways, which you already know of and use every day, to solve a problem :  

1. solve smaller problems, and combine the solution to solve your problem
2. solve a bigger problem, and specialize the solution to your problem

The first method is linked to the analytical process, of which, the breaking of an application into components is a result.

The second method involve genericity. Instead of an algorithm to sort two numbers, you produce an algorithm that sorts two **things**. The latter algorithm is more generic, not more complex. In fact, generic solutions are often precisely simpler (i.e. less interleaving things), in the sense of the talk you are referring to. They may feel however harder, in the sense of that same talk (easy = near to our capabilities), because of the lack of reference to concrete entities (the generic solution to composition that is a monad is a typical example, it is hard to grasp for some, but it is not complex) that we are used to manipulate and reason about.

I wrote about the benefits of componentization to tackle complexity on my blog [here](https://brucou.github.io/posts/componentization-against-complexity/). You will find the notions you mention about essential and incidental complexity, and the same notions that in the talk you mention.

Back to your remarks and question  :  

- obtuseness of composition : precisely, chips and circuits compose by design in a systematic, hierarchical and simple way. That is exactly what it is designed to be. It is a component model where the composition is port-based. It is generic enough to accomodate composition of things with things instead of just views with I don't really know what else. That is why you say obtuseness of composition for `cyclow`. It could be just a lack of documentation, and conceptual explanation on the side of the library author, but after reading examples code, I still don't see how those things compose. Composition being the primary pattern in functional programming, that is pretty annoying to me.
- yes, the architecture enables to draw the application as a tree or graphs of interconnected components. This is not to be underestimated as this is a potentially huge boon to maintanability and readability. It is not at all about the beautiful figure
- another benefit of the architecture is to force to design an interface before writing an implementation (cyclow has the same concept of ports, so that benefit also applies to cyclow). Just doing that eliminates bugs.
- the readouts and simulate that you see in the quick presentation of my design tackle exclusively the testing concern. Testing and debugging are key concerns to me. But  readouts can disappear from the picture if testing is not a concern. Testing in `cyclow` is pretty primitive and has not been developed further the use of `console`.
- the controller tackles cross-cutting concerns. i.e. concerns applying to the whole component tree.

**Comment by mangr3n** [comment]
I do plan to improve the API.  Graflow is the fundamental piece that makes sense to me.
I rapidly built a couple of graflow components to handle side effects (xhr,cookies).  Cyclow to me feels like AN attempt to implement components that draw on the ui using graflow under the covers.  For cyclow to succeed, to me I feel the need to be able to use the graflow API (send/on) directly on the components that are loaded into the browser, AND to be able to load them dynamically.  All of my code will not be cyclow components at the beginning and a migration path needs to be available to me.

There are many complex interactions between my state object, my server call, my localStorage interaction, and the ui components on the page that needs to be decomposed and abstracted before I put them back together.  Much of this will require some hammock time.

I've used redux and snabbdom to create a single directional flow model for my page.  The problem is I'm being asked to decompose my ui into reusable embeddable components.  I'm not going away from  state being a reduction on an event/action stream, or from view being a function of the current state.  But I didn't have a decomposition/composition story that allowed me to isolate my ui components for embedding and reuse.  graflow feels like a simple model for isolating system components

I've forked the codebases and will be checking in some changes.  Maybe we can continue this there.

**Comment by brucou** [comment]
Sure, send me the link for that repository. Your problem seem pretty common. An embeddable component receives events (from the user, or from the external systems, say the server, or local storage or else), and as a result may update its state and/or perform actions. Those actions in turn communicate back to the component by events. 

There are two common issues occurring here : concurrency, and control flow. 

Concurrency : a user clicks, then action is executed, before the action finishes he can click again and the same action may be executed again while it should not. Or a user can perform two actions which update the same remote table, but those updates might not occur in the same order the user intended it too, and you have the table in a wrong state. There are plenty of concurrency problems which can happen if you don't program against it.
Control flow : An action may also fail (server cannot fetch a resource for example), and the ui must be updated accordingly, possibly cancelling in-flight actions. Or some actions are only possible in some states, but the ui is designed so that those actions can be triggered by the user when they should not, and you need to control that programmatically. But if you dont do carefully you end up with if/then/else which is not very readable, error-prone and harder to maintain, after a while.

Those two issues are sources of complexity and bugs and is the hard part of ui programming. This is where good abstractions help a lot. Redux helps to manage shared state and separating action from action execution helps a bit for control flow (you can receive and discard based on some state in your store). It does not help so much for concurrency. React-saga (or just sagas if you don't use React) is a good complement for managing concurrency.

What will be interesting is to have a specification for one of your components, i.e. a function `action = f(state, event)`, and a function `next_state = f(action, state)` to understand the interaction between event, state, and actions. Of course, I am not talking about actually writing a function, but the words which explains the function. And from see what is an API that is generic enough to cover your needs

**Comment by mangr3n** [comment]
https://github.com/mangr3n/cyclow
https://github.com/mangr3n/graflow

**Comment by mangr3n** [comment]
So just to let you know, I've started making changes
* Add runComponent to get a reference to a component
* Added support for more Snabbdom features, props, attrs, style to the SnabbdomRenderer
* Modified the Component created by run/runComponent to put anything sent to out.signal in the on block to the returned component's out port.
